cmd/juju/commands/bootstrap.go

```diff
@@ -750,36 +757,64 @@ See `[1:] + "`juju kill-controller`" + `.`)
 	if c.AgentVersion != nil {
 		agentVersion = *c.AgentVersion
 	}
-	var addrs []network.Address
-	if env, ok := environ.(environs.InstanceBroker); ok {
-		// IAAS.
-		addrs, err = common.BootstrapEndpointAddresses(env, cloudCallCtx)
-		if err != nil {
-			return errors.Trace(err)
-		}
-	} else if env, ok := environ.(caas.ServiceGetterSetter); ok {
-		// CAAS.
-		var svc *caas.Service
-		svc, err = env.GetService(k8sprovider.JujuControllerStackName)
-		if err != nil {
-			return errors.Trace(err)
+
+	controllerDataRefresher := func() error {
+		// this function allows polling address info later during retring.
+		// for example, the Load Balancer needs time to be provisioned.
+		var addrs []network.Address
+		if env, ok := environ.(environs.InstanceBroker); ok {
+			// IAAS.
+			addrs, err = common.BootstrapEndpointAddresses(env, cloudCallCtx)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		} else if env, ok := environ.(caas.ServiceGetterSetter); ok {
+			// CAAS.
+			var svc *caas.Service
+			svc, err = env.GetService(k8sprovider.JujuControllerStackName)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if len(svc.Addresses) == 0 {
+				return errors.NotProvisionedf("k8s controller service %q address", svc.Id)
+			}
+			addrs = svc.Addresses
+		} else {
+			// TODO(caas): this should never happen.
+			return errors.NewNotValid(nil, "unexpected error happened, IAAS mode should have environs.Environ implemented.")
 		}
-		addrs = svc.Addresses
-	} else {
-		// TODO(caas): this should never happen.
-		return errors.NewNotValid(nil, "unexpected error happened, IAAS mode should have environs.Environ implemented.")
+
+		return errors.Annotate(
+			juju.UpdateControllerDetailsFromLogin(
+				c.ClientStore(),
+				c.controllerName,
+				juju.UpdateControllerParams{
+					AgentVersion:           agentVersion.String(),
+					CurrentHostPorts:       [][]network.HostPort{network.AddressesWithPort(addrs, config.controller.APIPort())},
+					PublicDNSName:          newStringIfNonEmpty(config.controller.AutocertDNSName()),
+					MachineCount:           newInt(1),
+					ControllerMachineCount: newInt(1),
+				},
+			),
+			"saving bootstrap endpoint address",
+		)
 	}
-	if err := juju.UpdateControllerDetailsFromLogin(
-		c.ClientStore(),
-		c.controllerName,
-		juju.UpdateControllerParams{
-			AgentVersion:           agentVersion.String(),
-			CurrentHostPorts:       [][]network.HostPort{network.AddressesWithPort(addrs, config.controller.APIPort())},
-			PublicDNSName:          newStringIfNonEmpty(config.controller.AutocertDNSName()),
-			MachineCount:           newInt(1),
-			ControllerMachineCount: newInt(1),
-		}); err != nil {
-		return errors.Annotate(err, "saving bootstrap endpoint address")
+
+	retryCallArgs := retry.CallArgs{
+		Attempts:    20,
```

Comment:
we already have some retry logic for IAAS controllers because the time it takes to boot a machine can vary wildly, especially with MAAS. And we already have Config to allow users to set those things.
Setting a max retry timeout of 1minute seems very aggressive (our default timeout is 10min)
Is there a reason we need yet-another retry loop here that doesn't use our existing timeout configuration?

Reply:
ah, I have moved this retry into `provider.Bootstrap()`. I probably should remove this retry here.
The reason why we  need the 2nd retry strategy for k8s controller is we need to ensure LB provisioning has been finished (LB address is ready to use) and we need the LB address in bootstrap-state cmd or the cmd will fail.
So we wait and fail in juju CLI if we can't get the address to avoid failing in bootstrap-state in Jujud.